### ISSUE TITLE
Add Sourcetrail to openSUSE repos

### ISSUE THREAD

**Issue by jubalh** [CoatiSoftware/Sourcetrail#758]
Similar to https://github.com/CoatiSoftware/Sourcetrail/issues/756 but for openSUSE.

I started working on this [here](https://build.opensuse.org/package/show/home:jubalh/sourcetrail) and will document progress here. And also mention once it landed in official repositories so it will be available via `zypper install sourcetree`.

**Comment by jubalh** [comment]
I get:

```
[   67s] CMake Error at /usr/lib64/cmake/Boost-1.71.0/BoostConfig.cmake:117 (find_package):
[   67s]   Found package configuration file:
[   67s] 
[   67s]     /usr/lib64/cmake/boost_system-1.71.0/boost_system-config.cmake
[   67s] 
[   67s]   but it set boost_system_FOUND to FALSE so package "boost_system" is
[   67s]   considered to be NOT FOUND.  Reason given by package:
[   67s] 
[   67s]   No suitable build variant has been found.
[   67s] 
[   67s]   The following variants have been tried and rejected:
[   67s] 
[   67s]   * libboost_system.so.1.71.0 (shared, Boost_USE_STATIC_LIBS=ON)
[   67s] 
```

Even though running cmake with: `-DBoost_USE_STATIC_LIBS=OFF`.

If I understand the current CMake script correctly then it's not possible to set this value to OFF via commandline right now.

Many distributions allow only shared libraries from their repos/package manager and not 3rd party shipped.

**Comment by jubalh** [comment]
I'll create an PR to make it a cache variable if that's okay.

**Comment by egraether** [comment]
I accepted the pull request. However we are wondering what you want to do? Do you want to compile, package and upload Sourcetrail yourself to openSUSE?

**Comment by jubalh** [comment]
> However we are wondering what you want to do? Do you want to compile, package and upload Sourcetrail yourself to openSUSE?

That's right.

**Comment by egraether** [comment]
> > However we are wondering what you want to do? Do you want to compile, package and upload Sourcetrail yourself to openSUSE?
> 
> That's right.

I see, we are no Linux experts. A couple infos on how we did things so far:
* We build our Linux release builds in docker containers based on CentOS6, which produce builds that are backwards compatible and work on all distributions (see `deployment/dockerfiles`). I don't know if this will be necessary for a openSUSE build.
* We currently package our Linux builds with CPack, but we are not happy about the solution and would like to switch to something else (e.g. flatpak, AppImage), but we don't really know what's best.
* We already added a tag for the next release version `2020.1` for all new developments from now on, which will be released in January or February. If we make changes to an already released version, we usually `cherry-pick` necessary changes from master onto a certain release branch (e.g. release_2019_4). If you want to upload a build publicly, please sync up with us upfront. We want to make sure that we know what's in this build and have it versioned `2019.4.XX`.

**Comment by jubalh** [comment]
Thanks for the infos.

> would like to switch to something else (e.g. flatpak, AppImage), but we don't really know what's best.

Yes the usual discussion. I really can't say anything about this since there are pros and cons of both sides. I tend to like AppImage more though (just my personal opinion).

> If you want to upload a build publicly, please sync up with us upfront.

Sure. That's why I created this 'issue'. To let you know that I plan to do this, discuss possible ways and inform others so work isn't duplicated.

**Comment by jubalh** [comment]
Currently I'm getting fails at linker level:

```
[ 1307s] /home/abuild/rpmbuild/BUILD/Sourcetrail-2019.4.61/src/lib_gui/utility/utilityApp.cpp:62: undefined reference to `FilePath::empty() const'
[ 1307s] /usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/Sourcetrail-2019.4.61/src/lib_gui/utility/utilityApp.cpp:64: undefined reference to `FilePath::wstr[abi:cxx11]() const'
[ 1307s] /usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: CMakeFiles/Sourcetrail_lib_gui.dir/src/lib_gui/utility/utilityApp.cpp.o: in function `utility::executeProcessUntilNoOutput(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, FilePath const&, int)':
[ 1307s] /home/abuild/rpmbuild/BUILD/Sourcetrail-2019.4.61/src/lib_gui/utility/utilityApp.cpp:98: undefined reference to `FilePath::empty() const'
[ 1307s] /usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/Sourcetrail-2019.4.61/src/lib_gui/utility/utilityApp.cpp:100: undefined reference to `FilePath::wstr[abi:cxx11]() const'
[ 1307s] /usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/Sourcetrail-2019.4.61/src/lib_gui/utility/utilityApp.cpp:120: undefined reference to `LogManager::getInstance()'
[ 1307s] /usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/Sourcetrail-2019.4.61/src/lib_gui/utility/utilityApp.cpp:120: undefined reference to `LogManager::logWarning(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int)'
[ 1307s] /usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: CMakeFiles/Sourcetrail_lib_gui.dir/src/lib_gui/utility/utilityApp.cpp.o: in function `utility::executeProcessAndGetExitCode(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::vector<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::allocator<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > > > const&, FilePath const&, int, bool, std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >*)':
[ 1307s] /home/abuild/rpmbuild/BUILD/Sourcetrail-2019.4.61/src/lib_gui/utility/utilityApp.cpp:188: undefined reference to `FilePath::empty() const'
[ 1307s] /usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/Sourcetrail-2019.4.61/src/lib_gui/utility/utilityApp.cpp:190: undefined reference to `FilePath::wstr[abi:cxx11]() const'
[ 1307s] /usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: CMakeFiles/Sourcetrail_lib_gui.dir/src/lib_gui/utility/path_detector/PathDetector.cpp.o: in function `void std::_Destroy<FilePath>(FilePath*)':
[ 1307s] /usr/include/c++/9/bits/stl_construct.h:98: undefined reference to `FilePath::~FilePath()'
[ 1307s] collect2: error: ld returned 1 exit status
```

[full build log](https://build.opensuse.org/build/home:jubalh/openSUSE_Tumbleweed/x86_64/sourcetrail/_log).

Is there a minimum/maximum Qt version requirement?

**Comment by egraether** [comment]
Did you pull the changes from #755 already?

**Comment by jubalh** [comment]
> Did you pull the changes from #755 already?

Yes, that patch is included. It's in my [repo](https://build.opensuse.org/package/show/home:jubalh/sourcetrail). And in the [spec](https://build.opensuse.org/package/view_file/home:jubalh/sourcetrail/sourcetrail.spec?expand=1) file one can read which patch is about which PR.

I see that in the CMake file we do `target_link_libraries(${LIB_UTILITY_PROJECT_NAME} ${Boost_LIBRARIES} Qt5::Widgets Qt5::Network)`. But it seems like linking to some boos functions doesnt work in my case. I don't see why.

[Full log](https://build.opensuse.org/build/home:jubalh/openSUSE_Tumbleweed/i586/sourcetrail/_log)

**Comment by jubalh** [comment]
https://github.com/CoatiSoftware/Sourcetrail/pull/784 is also needed.

**Comment by jubalh** [comment]
I see

`set(BUILD_PYTHON_LANGUAGE_PACKAGE OFF CACHE BOOL "Add Python support to the Sourcetrail indexer.")`

so that seems to be disabled by default. But I get:

```
[  918s] CMake Error at cmake_install.cmake:45 (file):
[  918s]   file INSTALL cannot find
[  918s]   "/home/abuild/rpmbuild/BUILD/Sourcetrail-2019.4.61/bin/app/data/python/SourcetrailPythonIndexer". 
```

@egraether so it seems to try to install it even though we said it shouldn't build it?

**Comment by egraether** [comment]
@jubalh I don't understand why it would look for `SourcetrailPythonIndexer` then.

Anyways, to create a complete build you need to enable all of:
```
BUILD_CXX_LANGUAGE_PACKAGE
BUILD_JAVA_LANGUAGE_PACKAGE
BUILD_PYTHON_LANGUAGE_PACKAGE
```

**Comment by jubalh** [comment]
> I don't understand why it would look for SourcetrailPythonIndexer then.

https://github.com/CoatiSoftware/Sourcetrail/pull/800 should take care of it.

> Anyways, to create a complete build you need to enable all of:

Yep, I'm aware. Thanks though!

**Comment by jubalh** [comment]
I first build was succesful.

The sourcetrail package now moved to an official openSUSE devel project at [devel:tools](https://build.opensuse.org/package/show/devel:tools/sourcetrail).

More people can now contribute to it, and once everything is polished it will go into official openSUSE repository.

**Comment by jubalh** [comment]
Is there a way to make sourcetrail not try to install:

```
/usr/Sourcetrail/lib/libEGL.so.1
/usr/Sourcetrail/lib/libGL.so.1
/usr/Sourcetrail/lib/libX11-xcb.so.1
/usr/Sourcetrail/lib/libX11.so.6
/usr/Sourcetrail/lib/libXau.so.6
/usr/Sourcetrail/lib/libXext.so.6
/usr/Sourcetrail/lib/libXrender.so.1
/usr/Sourcetrail/lib/libdrm.so.2
/usr/Sourcetrail/lib/libgbm.so.1
/usr/Sourcetrail/lib/libxcb-dri2.so.0
/usr/Sourcetrail/lib/libxcb-glx.so.0
/usr/Sourcetrail/lib/libxcb-render.so.0
/usr/Sourcetrail/lib/libxcb-shape.so.0
/usr/Sourcetrail/lib/libxcb-xfixes.so.0
/usr/Sourcetrail/lib/libxcb.so.1
```

But use the system libraries for this?

**Comment by jubalh** [comment]
We build an openSUSE package now at https://build.opensuse.org/package/show/devel:tools/sourcetrail

But we will wait for a next release before we go further with it. So it's not in official repositories yet but users can install it on their own risk with:

```
zypper ar -f obs://devel:tools
zypper ref
zypper in sourcetrail
```

FHS should be implemented, and probably more things need to be improved until it can go to official repos. But I saw that many PRs have been accepted. So we will wait for the next release and see then.

Closing this issue.
